### PR TITLE
Add linear projection onto plane

### PR DESCRIPTION
**Pull request by simonge**
### Briefly, what does this PR introduce?
Final step before CI pipelines can be introduced to train the ML method for reconstructing the Low-Q2 scattered electron momentum. Vectors from both tagger modules are projected onto a common plane from which a neural network can carry out the effective magnetic propagation back to the vertex.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No
### Does this PR change default behavior?
Adds new collections to the default output.

### CODE REVIEW

File: src/factories/fardetectors/FarDetectorLinearProjection_factory.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Created by Simon Gardner
+// Subject to the terms in the LICENSE file found in the top-level directory.
+//
```
Comment (veprbl) [inline src/factories/fardetectors/FarDetectorLinearProjection_factory.h:3]:
```suggestion
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright (C) 2023, Simon Gardner
```

File: src/factories/fardetectors/FarDetectorLinearProjection_factory.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    ParameterRef<std::vector<float>>     plane_position  {this, "plane_position", config().plane_position };
+    ParameterRef<std::vector<float>>     plane_a         {this, "plane_a",        config().plane_a        };
+    ParameterRef<std::vector<float>>     plane_b         {this, "plane_b",        config().plane_b        };
+
+  public:
+    void Configure() {
+        m_algo = std::make_unique<AlgoT>(GetPrefix());
+        m_algo->applyConfig(config());
+        m_algo->init(logger());
```
Comment (veprbl) [inline src/factories/fardetectors/FarDetectorLinearProjection_factory.h:35]:
```suggestion
        m_algo->level((algorithms::LogLevel)logger()->level());
        m_algo->applyConfig(config());
        m_algo->init();
```

File: src/algorithms/fardetectors/FarDetectorLinearProjection.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    public:
+        FarDetectorLinearProjection(std::string_view name)
+            : FarDetectorLinearProjectionAlgorithm{name,
+                {"inputTrackSegments"},
+                {"outputTrackParameters"},
+                "Project track segments to a plane"} {}
+
+        /** One time initialization **/
+        void init(std::shared_ptr<spdlog::logger>& logger);
```
Comment (veprbl) [inline src/algorithms/fardetectors/FarDetectorLinearProjection.h:40]:
```suggestion
        void init() final;
```

File: src/algorithms/fardetectors/FarDetectorLinearProjection.cc [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+#include <iterator>
+#include <algorithm>
+#include <map>
+
+
+
+namespace eicrecon {
+
+
+    void FarDetectorLinearProjection::init(std::shared_ptr<spdlog::logger>& logger) {
```
Comment (veprbl) [inline src/algorithms/fardetectors/FarDetectorLinearProjection.cc:20]:
```suggestion
    void FarDetectorLinearProjection::init() {
```

File: src/algorithms/fardetectors/FarDetectorLinearProjection.cc [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+      m_log      = logger;
+
```
Comment (veprbl) [inline src/algorithms/fardetectors/FarDetectorLinearProjection.cc:23]:
```suggestion
```

File: src/algorithms/fardetectors/FarDetectorLinearProjection.cc [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        // Create track parameters edm4eic structure
+        // TODO - populate more of the fields correctly
+        std::int32_t type = 0;
+        // Surface ID not used in this context
+        std::uint64_t surface = 0;
+        // Plane Point
+        edm4hep::Vector2f loc(projectedPoint[0],projectedPoint[1]); //Temp unit transform
+        float theta = inputPoint.theta;//edm4eic::anglePolar(outVec);
+        float phi   = inputPoint.phi  ;//edm4eic::angleAzimuthal(outVec);
+        float qOverP;
```
Comment (veprbl) [inline src/algorithms/fardetectors/FarDetectorLinearProjection.cc:58]:
```suggestion
        float qOverP = 0.;
```

File: src/factories/fardetectors/FarDetectorLinearProjection_factory.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    using AlgoT = eicrecon::FarDetectorLinearProjection;
+private:
+    std::unique_ptr<AlgoT> m_algo;
+
+    PodioInput<edm4eic::TrackSegment>     m_hits_input    {this};
+    PodioOutput<edm4eic::TrackParameters> m_tracks_output {this};
+
+    ParameterRef<std::vector<float>>     plane_position  {this, "plane_position", config().plane_position };
+    ParameterRef<std::vector<float>>     plane_a         {this, "plane_a",        config().plane_a        };
+    ParameterRef<std::vector<float>>     plane_b         {this, "plane_b",        config().plane_b        };
```
Comment (veprbl) [inline src/factories/fardetectors/FarDetectorLinearProjection_factory.h:29]:
```suggestion
    ParameterRef<std::vector<float>>     plane_position  {this, "planePosition", config().plane_position };
    ParameterRef<std::vector<float>>     plane_a         {this, "planeA",        config().plane_a        };
    ParameterRef<std::vector<float>>     plane_b         {this, "planeB",        config().plane_b        };
```
camelCase

File: src/algorithms/fardetectors/FarDetectorLinearProjection.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+        /** Event by event processing **/
+        void process(const Input&, const Output&) const final;
+
+
+    private:
+        std::shared_ptr<spdlog::logger> m_log;
```
Comment (veprbl) [inline src/algorithms/fardetectors/FarDetectorLinearProjection.h:47]:
```suggestion
```
Not needed, logging is mixed in https://github.com/eic/algorithms/blob/main/core/include/algorithms/logger.h

**Review by veprbl — APPROVED** [review]

**Comment by simonge** [comment]
Please could this be approved again after the suggestions were included.

**Comment by simonge** [comment]
Previous request was a bit premature as I hadn't taken into account the changes in the fitting step. Now output looks correct to me.

**Review by veprbl — APPROVED** [review]

**Pull request merged**